### PR TITLE
Adding example that uses the DMSDK to write data to MarkLogic

### PR DESCRIPTION
**Pull request by sastafford**

### CODE REVIEW

File: examples/dmsdk/build.gradle [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    }
+}
+
+dependencies {
+    compile "com.marklogic:java-client-api:version:4.0.0-EA2"
+    compile "com.marklogic:data-movement:1.0.0-EA2"
+    compile "com.marklogic:marklogic-spring-batch-core:0.5.2"
+    compile "com.marklogic:ml-javaclient-util:2.9.1"
+    testCompile "com.marklogic:marklogic-spring-batch-test:0.5.1"
+}
```
Comment (sastafford) [inline examples/dmsdk/build.gradle]: Here we are combining use of the MarkLogic Spring Batch library and DMSDK library from maven.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                batcher.onBatchFailure((client, batch, throwable) -> {
+                    System.err.println(throwable.toString());
+                });
+                manager.startJob(batcher);
+                logger.info("Size: " + items.size());
+                for (FileHandle item : items) {
+                    DocumentMetadataHandle meta = new DocumentMetadataHandle().withCollections(collections);
+                    batcher.add(UUID.randomUUID().toString(), meta, item);
+                }
+                batcher.awaitCompletion(1000, TimeUnit.MILLISECONDS);
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: The verification of the unit test was executing before the batcher finished writing content to MarkLogic.  So I'm using this awaitCompletion() function to put in a pause.  Only writing three files in the unit test.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.springframework.context.EnvironmentAware;
+import org.springframework.context.annotation.Bean;
+import org.springframework.core.env.Environment;
+import org.springframework.core.io.Resource;
+
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+public class ImportDocumentsWithDataMovementConfig extends AbstractMarkLogicBatchConfig implements EnvironmentAware {
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: The purpose of this Spring Batch job is to read some xml files from the file system and then push into MarkLogic.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    }
+    
+    @Bean
+    @JobScope
+    public Step step(
+            @Value("#{jobParameters['output_collections']}") String[] collections,
+            @Value("#{jobParameters['input_file_path']}") String inputFilePath,
+            @Value("#{jobParameters['input_file_pattern']}") String inputFilePattern) {
+    
+        ItemProcessor<Resource, FileHandle> processor = new MarkLogicImportItemProcessor();
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: This [MarkLogicImportItemProcessor](https://github.com/sastafford/marklogic-spring-batch/blob/dev/core/src/main/java/com/marklogic/spring/batch/item/MarkLogicImportItemProcessor.java) is an [ItemProcessor](http://docs.spring.io/spring-batch/apidocs/org/springframework/batch/item/ItemProcessor.html) that converts a Spring [Resource](http://docs.spring.io/spring-framework/docs/4.3.2.RELEASE/javadoc-api/org/springframework/core/io/Resource.html) into a MarkLogic [FileHandle](http://docs.marklogic.com/javadoc/client/com/marklogic/client/io/FileHandle.html).

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        //DMSDK Initialization
+        final DataMovementManager manager = DataMovementManager.newInstance();
+        manager.setClient(getDatabaseClient());
+        WriteHostBatcher batcher = manager.newWriteHostBatcher();
+    
+        ItemWriter<FileHandle> writer = new ItemWriter<FileHandle>() {
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: For simplicity, I just created an anonymous class to show how an ItemWriter could be created that leveraged the DMSDK.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    public Job job(Step step) {
+        return jobBuilderFactory.get(JOB_NAME).start(step).build();
+    }
+    
+    @Bean
+    @JobScope
+    public Step step(
+            @Value("#{jobParameters['output_collections']}") String[] collections,
+            @Value("#{jobParameters['input_file_path']}") String inputFilePath,
+            @Value("#{jobParameters['input_file_pattern']}") String inputFilePattern) {
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: These are the parameters that can be passed into the Job from a command line utility.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                batcher.awaitCompletion(1000, TimeUnit.MILLISECONDS);
+            }
+        };
+                   
+        return stepBuilderFactory.get("step1")
+                .<Resource, FileHandle>chunk(getChunkSize())
+                .reader(new EnhancedResourcesItemReader(inputFilePath, inputFilePattern))
+                .processor(processor)
+                .writer(writer)
+                .build();
```
Comment (sastafford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: This is the magic step that connects up the ItemReader, ItemProcessor, and ItemWriter into a single step.  Note the getChunkSize() function.  That is from the AbstractMarkLogicBatchConfig class.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+                });
+                batcher.onBatchFailure((client, batch, throwable) -> {
+                    System.err.println(throwable.toString());
+                });
+                manager.startJob(batcher);
+                logger.info("Size: " + items.size());
+                for (FileHandle item : items) {
+                    DocumentMetadataHandle meta = new DocumentMetadataHandle().withCollections(collections);
+                    batcher.add(UUID.randomUUID().toString(), meta, item);
+                }
```
Comment (sammefford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: Here you need batcher.flush() or there might be a batch that isn't full that never gets sent to ML server.

File: examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                logger.info("DMSDK ItemWriter");
+                batcher.withJobName(JOB_NAME);
+                batcher.withBatchSize(items.size());
+                batcher.onBatchSuccess((client, batch) -> {
+                    System.out.println("Sucessfully wrote " + batch.getItems().length);
+                });
+                batcher.onBatchFailure((client, batch, throwable) -> {
+                    System.err.println(throwable.toString());
+                });
+                manager.startJob(batcher);
```
Comment (sammefford) [inline examples/dmsdk/src/main/java/example/ImportDocumentsWithDataMovementConfig.java]: would it make sense to put more of the batcher config outside this write method (including startJob)?  It probably doesn't need to be initialized per batch.  And startJob is intended to only be called once.

**Pull request merged**